### PR TITLE
[perf] ShardedDDP: better handling of the callback queue, try to consume it as we go.

### LINKED ISSUE #253: [ShardedDDP] Reduce the gradient bubble
## 🚀 Feature
Right now the backward pass is allowed to proceed all the way before the gradients are cleaned up, which means that all the gradients materialize. The peak memory use is bigger than what it should be if we clean them on the fly as the BW pass proceeds

## Motivation
Free more memory

## Pitch
See attached PR, look up the futures dequeue when firing up a BW hook, and cleanup whatever is ready (non blocking, important for perf)

## Alternatives
Not doing that, works but memory waste

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #253

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blefaudeux** [comment]
(lint errors unrelated, being fixed in another PR)

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 1]
```diff
@@ -111,6 +113,7 @@ def forward(self, *inputs: Any, **kwargs: Any) -> Any:
 
         # Reset all the grad reduce and bucket state flags
         self._grad_to_be_reduced = [True] * len(self._grad_to_be_reduced)
+        self._current_reduce_calls = {o: 0 for o in self.sharded_optimizers}
```
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/sharded_ddp.py:116]: put this in a function to reduce duplication with line 95?

File: fairscale/optim/oss.py [hunk 2]
```diff
@@ -107,6 +104,11 @@ def __init__(
         # Current default device is set by the parameters allocated to this rank
         self._device = list(self.per_device_params.keys())[0]
         self.buckets: Dict[torch.device, List[Bucket]] = {}
+
+        if self.world_size <= 8:
```
Comment (min-xu-ai) [inline fairscale/optim/oss.py:108]: you can use torch.cuda.get_device_count()?

File: pyproject.toml [hunk 3]
```diff
@@ -28,4 +28,4 @@ use_parentheses = true
 skip_glob = ["build/*", "stubs/*"]
 # Don't split "import" and "from".
 force_sort_within_sections = true
-known_third_party = ["benchmark_dataset", "dataclasses", "numpy", "pytest", "recommonmark", "setuptools", "torch", "torch_pg", "torchtext", "torchvision"]
+known_third_party = ["benchmark_dataset", "dataclasses", "helpers", "numpy", "pytest", "recommonmark", "setuptools", "torch", "torch_pg", "torchtext", "torchvision"]
```
Comment (min-xu-ai) [inline pyproject.toml:31]: where is helpers from? I didn't see it in this PR.

**Review by min-xu-ai — COMMENTED** [review]
some minor comments. would love to not hardcode 8.

File: fairscale/optim/oss.py (hunk 2 above)
Comment (blefaudeux) [reply fairscale/optim/oss.py:108]: so ` if self.world_size == torch.cuda.get_device_count()` really ? yeah, makes sense

File: pyproject.toml (hunk 3 above)
Comment (blefaudeux) [reply pyproject.toml:31]: oh, I've absolutely no idea where that came from VSCode shenanigan I suppose, removing that

**Comment by blefaudeux** [comment]
In practice, 300MB saved on a .8B parameters model / 8 gpus.

**Review by min-xu-ai — APPROVED** [review]

File: pyproject.toml (hunk 3 above)
Comment (blefaudeux) [reply pyproject.toml:31]: the git commit hooks are the ones creating this file, not how that works to be honest

**Comment by joshim5** [comment]
This looks awesome! Have you profiled if there is any impact on speed?

**Comment by blefaudeux** [comment]
> This looks awesome! Have you profiled if there is any impact on speed?

yes I have :) (FB internal https://docs.google.com/spreadsheets/d/1JPxxhbN60G0eHNSDeMNf3q-mPkjBnGtfZGT96t_p_PU/edit?usp=sharing), this is the v3 basically (some of them). Not done yet testing the whole landscape, but if anything it's a bit faster because this futures consumption overlaps a bit more with the bw computations

**Pull request merged**